### ISSUE TITLE
memory: retrieval gate should be able to skip injection entirely, not just filter per-result

### ISSUE THREAD

**Issue by schancel** [darwin-finch/finch#1134]
Memory recall currently only filters per-result by a flat score floor; it never decides to inject nothing at all for a turn that doesn't need memory context.

## Current behavior

`MemoryConfig::min_relevance_score` (`crates/finch-memory/src/lib.rs:281`, default `0.15` at `:305`) is applied once per retrieved result in `query_with_sources` (`crates/finch-memory/src/lib.rs:1689-1698`):

```rust
let min_score = self.config.min_relevance_score;
...
for (node_id, text, score) in retrieved {
    if score < min_score {
        continue;
    }
    ...
}
```

That drops individual weak matches (#940), but it's a per-item cutoff, not a per-turn decision. If a query needs no memory context at all, but the tree happens to return `k` results that each individually clear `0.15`, they still get injected -- there's no gate asking "should anything be injected this turn."

## What's wanted

A smarter turn-level gate: decide whether to inject anything at all before falling back to per-result filtering, rather than relying on a single flat floor to do both jobs.

## Possible prior art -- needs confirmation

There was reportedly an earlier eval/judge-style setup using Qwen where the model itself could refuse to answer rather than being forced to respond. If that carries lessons for how a turn-level "don't inject" decision should be made (e.g. what signal it used to decide to abstain, how it was evaluated), it's worth pulling in -- but I couldn't find it in this repo's history or docs, so whoever picks this up should confirm where that work lives (a branch, a doc, the sibling `fractal-corpus-curation` research repo) before assuming it transfers directly.

## Not yet done

- No design proposed yet for what signal the turn-level gate should use (query intent classification? a confidence/margin check on the best result? something else?).
- No threshold calibration against real usage has happened for the existing `min_relevance_score` floor either -- that's a related, separate gap.

**Comment by schancel** [comment]
Update from Shammah on the prior-art question above: the earlier system used a logit-based approach -- an LLM (Qwen) generated a summary for each candidate memory node, and the routing/inclusion decision was made from that model's own output logits rather than embedding cosine similarity, with the model able to refuse/decline rather than being forced into a binary choice. That's a materially different signal from the current flat `min_relevance_score` cosine floor: a learned relevance judgment produced by generation itself, not a distance metric.

I still haven't located this system in this repo (its code, if committed, may live elsewhere -- possibly the sibling `fractal-corpus-curation` research repo referenced in `crates/finch-memory/AGENTS.md` for the routing tree's own D-reference port). One related, but not identical, structural precedent already in this repo: `docs/TRAINING_FRAMEWORK.md:411-433` describes a "Router: Binary Cross-Entropy" design -- a single logit through `sigmoid` producing a forward-probability -- but that's for the local-vs-cloud query router, not memory-node summary routing, and per this repo's own caveats (`CLAUDE.md`, local training is deferred/unimplemented) it's a design doc, not evidence the mechanism was ever built or trained here.

Whoever picks this up should track down the actual Qwen/logit memory-routing system (ask Shammah directly for where it lives) before designing the turn-level gate, rather than reinventing it from this description alone.

**Comment by schancel** [comment]
Correction: Shammah doesn't think it was in `fractal-corpus-curation` either, and doesn't recall where it lives. Treat the location as genuinely unknown -- don't chase that repo as a lead.

**Comment by schancel** [comment]
**Compact solution contract — turn-level retrieval skip gate (#1134)**

**Accepted outcome.** Memory recall gains a turn-level decision to inject *nothing at all* for a turn that does not need memory context, made before the per-result `min_relevance_score` floor is applied. Today `query_with_sources` (`crates/finch-memory/src/lib.rs:1876` at base 81bfdafa) only filters per result (default floor `0.15`), and nearest-neighbour retrieval always returns `k` candidates — so a query needing no memory context still pays for `k` weak-but-above-floor memories injected into every turn.

**Gate condition and signal.** A confidence check on the best retrieved result: when the highest weighted score (cosine × importance boost, the same quantity the per-result floor uses) among this turn's retrieved candidates is strictly below a configured turn floor, the turn injects nothing. Otherwise the existing per-result floor applies unchanged. This is the ticket's "confidence/margin check on the best result" option, chosen because it is deterministic, cheap, and testable. The ticket's other candidate signals are explicit non-goals here: no query-intent classifier, and no LLM/logit signal — the Qwen logit prior art is unlocated per the issue's own comments, and the maintainer should rule on it before any learned signal replaces this one.

**Config semantics.** New `MemoryConfig::min_turn_relevance_score: Option<f32>`, default `None` — the gate is **OFF by default**, so behavior is identical to today until it is enabled. `Some(t)` enables it. Documented relationship: `t` only changes behavior when it is strictly above `min_relevance_score` (otherwise the per-result floor already empties those turns). Like the existing floor, `t` is an uncalibrated starting point — threshold calibration against real usage is the ticket's separately-tracked gap and is not done here. TOML/CLI wiring for the knob touches `src/cli/**`, which is out of this change's allowed scope; the knob is reachable by anyone constructing `MemoryConfig` (tests, runtime) and the config.toml wiring is a named follow-up.

**Observability.** The gate's decision is logged: a skip is logged at `info` with the turn floor, best score, and candidate count; an allow under an enabled gate is logged at `debug`; the per-result floor's drop count joins the existing `debug` result line. The decision is computed by one pure function, unit-tested, and the log payload carries exactly the tested values.

**Invariants kept.** The per-result floor still applies whenever injection happens; retrieval ordering and the retrieval algorithm are untouched (the retrieval-algorithm lane belongs to #948, routing-tree work); default OFF means no change for existing users.

**Interaction disclosed, not changed.** A skipped turn returns an empty recall set, which the committed-memory-set aging in the query processor (`stale_after_turns`) will count as a turn without reappearing — with the gate OFF this never happens; with it ON, committed memories age on gated turns. That logic lives in `src/cli/**` (out of scope) and is disclosed here for the reviewer.

**Overlap with the in-flight memory lane — flagged, no conflict found.** Open PR #1125 (RoutingTree replaces MemTree; retrieval-time raw-vs-summarize gate) gates the *presentation of individual entries* (raw vs extractive summary); it does not decide whether anything is injected for a turn, so #1134's outcome is distinct and unimplemented there. Merge-order note: that PR rewrites this crate; if it lands first, this gate re-lands on the RoutingTree-backed `query_with_sources` (the facade survives there). Maintainer ruling welcome on (a) eventual signal and (b) default OFF vs ON; this PR ships the deterministic floor variant, OFF, as the minimal deliverable the ticket asks for.

**Proof.** Fail-before/pass-after regressions in the crate: gate skips a turn whose best score previously injected (and whose best score cleared the per-result floor — the leak precondition is asserted); gate allows strong results; per-result floor still drops weak matches on allowed turns; the knob flips the decision between two identically-seeded stores; default-OFF behavior matches baseline; decision-function mapping including the strictly-below boundary and empty-retrieval case; the skip/allow log lines are captured and asserted. `finch-memory --lib` and the supervised workspace lib suite green; `cargo fmt`; no new clippy warnings.

**Rollback.** The knob is `Option`, default `None`: setting it back to `None` (or reverting the commit) restores exact prior behavior. No schema, storage, or wire-format change.

**Comment by schancel** [comment]
`opencode` worker (@schancel) is claiming implementation of #1134 for a turn-level injection skip gate in `crates/finch-memory` (`query_with_sources`, `MemoryConfig` knob, tests, capsule docs). Lane: primary harness model only — this work must not be re-dispatched to a cheaper model. No conflict with open claims observed; PR #1125's per-entry summarize gate does not implement this ticket's turn-level skip and is flagged (not blocked) in the solution contract above.

**Comment by schancel** [comment]
Correction to the claim above (posted seconds ago): the `base:` field carried a mistyped full SHA. The correct full base commit for worktree `/Users/shammah/repos/finch-issue-1134` on branch `fix/1134-retrieval-skip-gate` is `81bfdafa4aed98eb470ded8eb7cd3d5c6ff20488`. Same claim, same scope; this block supersedes the base field of the previous claim block.

**Comment by schancel** [comment]
Implemented by PR #1138: the retrieval gate can skip injection entirely (config-driven, observable skip/allow decision lines, per-result floor preserved). Maintainer rulings requested: #1125 merge order, the eventual recall signal replacing the score placeholder, and default OFF vs ON. Claim released.
